File: go/internal/guard/rules_test.go
package guard

import (
	"testing"

	"github.com/Chachamaru127/claude-code-harness/go/pkg/protocol"
)

// helper to build a RuleContext for testing
func makeCtx(toolName string, toolInput map[string]interface{}) protocol.RuleContext {
	return protocol.RuleContext{
		Input: protocol.HookInput{
			ToolName:  toolName,
			ToolInput: toolInput,
		},
		ProjectRoot:  "/project",
		WorkMode:     false,
		CodexMode:    false,
		BreezingRole: "",
	}
}

// ---------------------------------------------------------------------------
// R01: sudo block
// ---------------------------------------------------------------------------

func TestR01_SudoBlocked(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "sudo rm -rf /"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR01_SudoInMiddle(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "echo hello && sudo apt install"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR01_NoSudo(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "ls -la"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R02: protected path write block
// ---------------------------------------------------------------------------

func TestR02_WriteToEnv(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": ".env"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR02_WriteToGitDir(t *testing.T) {
	ctx := makeCtx("Edit", map[string]interface{}{"file_path": ".git/config"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR02_WriteToIdRsa(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/home/user/.ssh/id_rsa"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR02_WriteToNormalFile(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/src/main.ts"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve, got %s", result.Decision)
	}
}

func TestR02_WriteToPemFile(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "certs/server.pem"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R03: Bash write to protected paths
// ---------------------------------------------------------------------------

func TestR03_EchoToEnv(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "echo SECRET=foo > .env"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR03_TeeToGit(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "echo test | tee .git/config"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR03_NormalBash(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "echo hello > output.txt"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R04: write outside project root
// ---------------------------------------------------------------------------

func TestR04_WriteOutsideProject(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/tmp/malicious.sh"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionAsk {
		t.Errorf("expected ask, got %s", result.Decision)
	}
}

func TestR04_WriteInsideProject(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/src/index.ts"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve, got %s", result.Decision)
	}
}

func TestR04_RelativePath(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "src/index.ts"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve, got %s", result.Decision)
	}
}

func TestR04_WorkModeBypass(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/tmp/file.txt"})
	ctx.WorkMode = true
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve in work mode, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R05: rm -rf confirmation
// ---------------------------------------------------------------------------

func TestR05_RmRfBlocked(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "rm -rf /var/data"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionAsk {
		t.Errorf("expected ask, got %s", result.Decision)
	}
}

func TestR05_RmRfWorkMode(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "rm -rf ./dist"})
	ctx.WorkMode = true
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve in work mode, got %s", result.Decision)
	}
}

func TestR05_RmFOnly(t *testing.T) {
	// rm -f (without -r) should NOT trigger R05
	ctx := makeCtx("Bash", map[string]interface{}{"command": "rm -f temp.txt"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve for rm -f (no -r), got %s", result.Decision)
	}
}

func TestR05_RmRecursive(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "rm --recursive ./dir"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionAsk {
		t.Errorf("expected ask, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R06: force push block
// ---------------------------------------------------------------------------

func TestR06_ForcePush(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git push --force origin main"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR06_ForceWithLease(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git push --force-with-lease origin feature"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR06_ShortForce(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git push -f origin main"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR06_NormalPush(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git push origin feature"})
	result := EvaluateRules(ctx)
	// R12 might trigger a warning for protected branch, but this is "feature"
	if result.Decision == protocol.DecisionDeny {
		t.Errorf("expected non-deny for normal push, got deny")
	}
}

// ---------------------------------------------------------------------------
// R07: Codex mode write block
// ---------------------------------------------------------------------------

func TestR07_CodexModeWrite(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/src/main.ts"})
	ctx.CodexMode = true
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny in codex mode, got %s", result.Decision)
	}
}

func TestR07_CodexModeEdit(t *testing.T) {
	ctx := makeCtx("Edit", map[string]interface{}{"file_path": "/project/src/main.ts"})
	ctx.CodexMode = true
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny in codex mode, got %s", result.Decision)
	}
}

func TestR07_CodexModeBash(t *testing.T) {
	// Bash is NOT blocked by R07 (only Write/Edit/MultiEdit)
	ctx := makeCtx("Bash", map[string]interface{}{"command": "echo hello"})
	ctx.CodexMode = true
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve for Bash in codex mode, got %s", result.Decision)
	}
}

func TestR07_NormalModeWrite(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/src/main.ts"})
	ctx.CodexMode = false
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve in normal mode, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R08: Breezing reviewer write block
// ---------------------------------------------------------------------------

func TestR08_ReviewerWrite(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/src/main.ts"})
	ctx.BreezingRole = "reviewer"
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny for reviewer write, got %s", result.Decision)
	}
}

func TestR08_ReviewerBashGitCommit(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git commit -m 'test'"})
	ctx.BreezingRole = "reviewer"
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny for reviewer git commit, got %s", result.Decision)
	}
}

func TestR08_ReviewerBashReadOnly(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "cat README.md"})
	ctx.BreezingRole = "reviewer"
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve for reviewer read-only bash, got %s", result.Decision)
	}
}

func TestR08_WorkerWrite(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/src/main.ts"})
	ctx.BreezingRole = "worker"
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve for worker write, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R09: secret file read warning
// ---------------------------------------------------------------------------

func TestR09_ReadEnv(t *testing.T) {
	ctx := makeCtx("Read", map[string]interface{}{"file_path": "/project/.env"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve (warning only), got %s", result.Decision)
	}
	if result.SystemMessage == "" {
		t.Error("expected a warning systemMessage")
	}
}

func TestR09_ReadIdRsa(t *testing.T) {
	ctx := makeCtx("Read", map[string]interface{}{"file_path": "/home/user/.ssh/id_rsa"})
	result := EvaluateRules(ctx)
	if result.SystemMessage == "" {
		t.Error("expected a warning systemMessage for id_rsa")
	}
}

func TestR09_ReadNormalFile(t *testing.T) {
	ctx := makeCtx("Read", map[string]interface{}{"file_path": "/project/README.md"})
	result := EvaluateRules(ctx)
	if result.SystemMessage != "" {
		t.Errorf("expected no warning for normal file, got: %s", result.SystemMessage)
	}
}

// ---------------------------------------------------------------------------
// R10: --no-verify / --no-gpg-sign block
// ---------------------------------------------------------------------------

func TestR10_NoVerify(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git commit --no-verify -m 'test'"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR10_NoGpgSign(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git commit --no-gpg-sign -m 'test'"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR10_NormalCommit(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git commit -m 'test'"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R11: protected branch reset --hard
// ---------------------------------------------------------------------------

func TestR11_ResetHardMain(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git reset --hard origin/main"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR11_ResetHardMaster(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git reset --hard master"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestR11_ResetHardFeature(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git reset --hard origin/feature"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve for non-protected branch, got %s", result.Decision)
	}
}

func TestR11_ResetSoftMain(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git reset --soft main"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve for soft reset, got %s", result.Decision)
	}
}

// ---------------------------------------------------------------------------
// R12: direct push to protected branch warning
// ---------------------------------------------------------------------------

func TestR12_PushToMain(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git push origin main"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve (warning only), got %s", result.Decision)
	}
	if result.SystemMessage == "" {
		t.Error("expected warning systemMessage for push to main")
	}
}

func TestR12_PushToFeature(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git push origin feature-branch"})
	result := EvaluateRules(ctx)
	if result.SystemMessage != "" {
		t.Errorf("expected no warning for feature branch push, got: %s", result.SystemMessage)
	}
}

func TestR12_PushRefspecToMain(t *testing.T) {
	ctx := makeCtx("Bash", map[string]interface{}{"command": "git push origin HEAD:main"})
	result := EvaluateRules(ctx)
	if result.SystemMessage == "" {
		t.Error("expected warning for refspec push to main")
	}
}

// ---------------------------------------------------------------------------
// R13: protected review paths warning
// ---------------------------------------------------------------------------

func TestR13_WritePackageJson(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/package.json"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve (warning only), got %s", result.Decision)
	}
	if result.SystemMessage == "" {
		t.Error("expected warning systemMessage for package.json")
	}
}

func TestR13_WriteDockerfile(t *testing.T) {
	ctx := makeCtx("Edit", map[string]interface{}{"file_path": "Dockerfile"})
	result := EvaluateRules(ctx)
	if result.SystemMessage == "" {
		t.Error("expected warning for Dockerfile")
	}
}

func TestR13_WriteGitHubWorkflow(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": ".github/workflows/ci.yml"})
	result := EvaluateRules(ctx)
	if result.SystemMessage == "" {
		t.Error("expected warning for GitHub workflow")
	}
}

func TestR13_WriteNormalFile(t *testing.T) {
	ctx := makeCtx("Write", map[string]interface{}{"file_path": "/project/src/utils.ts"})
	result := EvaluateRules(ctx)
	if result.SystemMessage != "" {
		t.Errorf("expected no warning for normal file, got: %s", result.SystemMessage)
	}
}

// ---------------------------------------------------------------------------
// Rule evaluation order: first match wins
// ---------------------------------------------------------------------------

func TestFirstMatchWins(t *testing.T) {
	// sudo rm -rf should be caught by R01 (sudo) before R05 (rm -rf)
	ctx := makeCtx("Bash", map[string]interface{}{"command": "sudo rm -rf /"})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
	// R01 reason mentions sudo
	if result.Reason == "" || result.Reason[0:4] != "sudo" {
		// Check that the reason is about sudo, not rm -rf
		if result.Reason != "sudo の使用は禁止されています。必要な場合はユーザーに手動実行を依頼してください。" {
			t.Errorf("expected sudo reason, got: %s", result.Reason)
		}
	}
}

func TestUnknownToolApproved(t *testing.T) {
	ctx := makeCtx("UnknownTool", map[string]interface{}{})
	result := EvaluateRules(ctx)
	if result.Decision != protocol.DecisionApprove {
		t.Errorf("expected approve for unknown tool, got %s", result.Decision)
	}
}

File: core/src/guardrails/__tests__/rules.test.ts
/**
 * core/src/guardrails/__tests__/rules.test.ts
 * Unit tests for GUARD_RULES declarative guard rule table
 *
 * Verifies that each rule from pretooluse-guard.sh is correctly ported to TypeScript.
 * Coverage target: 90%+
 */

import { describe, it, expect } from "vitest";
import { GUARD_RULES, evaluateRules } from "../rules.js";
import type { RuleContext, HookInput } from "../../types.js";

// ============================================================
// Test helpers
// ============================================================

function makeCtx(
  toolName: string,
  toolInput: Record<string, unknown> = {},
  overrides: Partial<Omit<RuleContext, "input">> = {}
): RuleContext {
  const input: HookInput = { tool_name: toolName, tool_input: toolInput };
  return {
    input,
    projectRoot: "/project",
    workMode: false,
    codexMode: false,
    breezingRole: null,
    ...overrides,
  };
}

// ============================================================
// R01: sudo block
// ============================================================
describe("R01: sudo block", () => {
  it("blocks sudo rm -rf /", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "sudo rm -rf /" })
    );
    expect(result.decision).toBe("deny");
  });

  it("blocks sudo apt-get install", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "sudo apt-get install vim" })
    );
    expect(result.decision).toBe("deny");
  });

  it("does not block commands without sudo prefix", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "nosudo echo test" })
    );
    expect(result.decision).toBe("approve");
  });

  it("does not block Bash without sudo", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "ls -la" })
    );
    expect(result.decision).toBe("approve");
  });

  it("does not apply to Write tool", () => {
    // R01 targets Bash only
    const rule = GUARD_RULES.find((r) => r.id === "R01:no-sudo")!;
    const result = rule.evaluate(
      makeCtx("Write", { file_path: "/project/sudo.ts" })
    );
    expect(result).toBeNull();
  });
});

// ============================================================
// R02: protected path write block
// ============================================================
describe("R02: protected path write block", () => {
  const protectedPaths = [
    ".git/config",
    "/project/.git/HEAD",
    ".env",
    "/project/.env",
    "/home/user/.env.local",
    "credentials.pem",
    "private.key",
    "id_rsa",
    "id_ed25519",
    "/home/user/.ssh/id_ecdsa",
  ];

  for (const path of protectedPaths) {
    it(`blocks Write to ${path}`, () => {
      const result = evaluateRules(
        makeCtx("Write", { file_path: path })
      );
      expect(result.decision).toBe("deny");
    });

    it(`blocks Edit to ${path}`, () => {
      const result = evaluateRules(
        makeCtx("Edit", { file_path: path })
      );
      expect(result.decision).toBe("deny");
    });
  }

  it("does not block Write to normal source file", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "/project/src/index.ts" })
    );
    expect(result.decision).toBe("approve");
  });

  it("R02 does not apply to Bash tool", () => {
    const rule = GUARD_RULES.find((r) => r.id === "R02:no-write-protected-paths")!;
    const result = rule.evaluate(
      makeCtx("Bash", { command: "echo hello > .env" })
    );
    expect(result).toBeNull();
  });
});

// ============================================================
// R03: Bash shell write to protected paths block
// ============================================================
describe("R03: Bash shell write to protected paths block", () => {
  const dangerousBashCmds = [
    'echo "SECRET=foo" > .env',
    'echo "key" > .env.local',
    "cat token.txt > .git/config",
    "tee .git/hooks/pre-commit",
    "cat private.key > backup.key",
  ];

  for (const cmd of dangerousBashCmds) {
    it(`blocks ${cmd}`, () => {
      const result = evaluateRules(makeCtx("Bash", { command: cmd }));
      expect(result.decision).toBe("deny");
    });
  }

  it("does not block safe Bash commands", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "echo hello" })
    );
    expect(result.decision).toBe("approve");
  });
});

// ============================================================
// R04: write outside project confirmation
// ============================================================
describe("R04: write outside project confirmation", () => {
  it("returns ask for Write to absolute path outside project", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "/tmp/output.txt" }, { projectRoot: "/project" })
    );
    expect(result.decision).toBe("ask");
  });

  it("returns ask for Edit to absolute path outside project", () => {
    const result = evaluateRules(
      makeCtx("Edit", { file_path: "/home/user/outside.ts" }, { projectRoot: "/project" })
    );
    expect(result.decision).toBe("ask");
  });

  it("does not return ask for absolute path inside project", () => {
    const result = evaluateRules(
      makeCtx(
        "Write",
        { file_path: "/project/src/foo.ts" },
        { projectRoot: "/project" }
      )
    );
    expect(result.decision).toBe("approve");
  });

  it("treats relative paths as inside project", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "src/foo.ts" })
    );
    expect(result.decision).toBe("approve");
  });

  it("does not confirm outside-project writes in work mode", () => {
    const result = evaluateRules(
      makeCtx(
        "Write",
        { file_path: "/tmp/output.txt" },
        { workMode: true, projectRoot: "/project" }
      )
    );
    // R04 is skipped in workMode → subsequent rules approve
    expect(result.decision).toBe("approve");
  });
});

// ============================================================
// R05: rm -rf confirmation
// ============================================================
describe("R05: rm -rf confirmation", () => {
  const rmRfCmds = [
    "rm -rf /tmp/work",
    "rm -fr /tmp/work",
    "rm --recursive /tmp/work",
    "rm -rf ~/Downloads/old",
  ];

  for (const cmd of rmRfCmds) {
    it(`${cmd} returns ask`, () => {
      const result = evaluateRules(makeCtx("Bash", { command: cmd }));
      expect(result.decision).toBe("ask");
    });
  }

  it("does not confirm rm -rf in work mode", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "rm -rf /tmp/work" }, { workMode: true })
    );
    // R05 skipped in workMode → falls through to R06 (no match → approve)
    expect(result.decision).toBe("approve");
  });

  it("does not block normal rm -f", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "rm -f /tmp/test.log" })
    );
    expect(result.decision).toBe("approve");
  });
});

// ============================================================
// R06: git push --force block
// ============================================================
describe("R06: git push --force block", () => {
  const forcePushCmds = [
    "git push --force",
    "git push --force-with-lease",
    "git push origin main --force",
    "git push -f",
    "git push origin main -f",
  ];

  for (const cmd of forcePushCmds) {
    it(`blocks ${cmd}`, () => {
      const result = evaluateRules(makeCtx("Bash", { command: cmd }));
      expect(result.decision).toBe("deny");
    });
  }

  it("does not block normal git push", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git push origin feature/login" })
    );
    expect(result.decision).toBe("approve");
  });

  it("blocks force push even in work mode (no exceptions)", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git push --force" }, { workMode: true })
    );
    expect(result.decision).toBe("deny");
  });
});

// ============================================================
// R10: Git bypass flags block
// ============================================================
describe("R10: Git bypass flags block", () => {
  it("blocks --no-verify", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git commit --no-verify -m 'test'" })
    );
    expect(result.decision).toBe("deny");
  });

  it("blocks --no-gpg-sign", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git commit --no-gpg-sign -m 'test'" })
    );
    expect(result.decision).toBe("deny");
  });
});

// ============================================================
// R11: git reset --hard to protected branch block
// ============================================================
describe("R11: git reset --hard to protected branch block", () => {
  const dangerousResetCmds = [
    "git reset --hard main",
    "git reset --hard master",
    "git reset --hard origin/main",
  ];

  for (const cmd of dangerousResetCmds) {
    it(`blocks ${cmd}`, () => {
      const result = evaluateRules(makeCtx("Bash", { command: cmd }));
      expect(result.decision).toBe("deny");
    });
  }
});

// ============================================================
// R12: direct push to protected branch warning
// ============================================================
describe("R12: direct push to protected branch warning", () => {
  it("git push origin main returns approve + systemMessage", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git push origin main" })
    );
    expect(result.decision).toBe("approve");
    expect(result.systemMessage).toBeTruthy();
    expect(result.systemMessage).toContain("main");
  });

  it("git push upstream master returns approve + systemMessage", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git push upstream master" })
    );
    expect(result.decision).toBe("approve");
    expect(result.systemMessage).toBeTruthy();
    expect(result.systemMessage).toContain("master");
  });
});

// ============================================================
// R13: important file change warning
// ============================================================
describe("R13: important file change warning", () => {
  const protectedPaths = [
    "package.json",
    "Dockerfile",
    "docker-compose.yml",
    ".github/workflows/ci.yml",
    "schema.prisma",
    "wrangler.toml",
    "index.html",
  ];

  for (const path of protectedPaths) {
    it(`Write to ${path} returns approve + systemMessage`, () => {
      const result = evaluateRules(
        makeCtx("Write", { file_path: path })
      );
      expect(result.decision).toBe("approve");
      expect(result.systemMessage).toBeTruthy();
      expect(result.systemMessage).toContain(path);
    });
  }

  it("normal source file change does not warn", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "src/index.ts" })
    );
    expect(result.decision).toBe("approve");
    expect(result.systemMessage).toBeUndefined();
  });
});

// ============================================================
// R07: Write/Edit block in Codex mode
// ============================================================
describe("R07: Write/Edit block in Codex mode", () => {
  it("blocks Write in Codex mode", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "/project/src/foo.ts" }, { codexMode: true })
    );
    expect(result.decision).toBe("deny");
  });

  it("blocks Edit in Codex mode", () => {
    const result = evaluateRules(
      makeCtx("Edit", { file_path: "/project/src/foo.ts" }, { codexMode: true })
    );
    expect(result.decision).toBe("deny");
  });

  it("does not block Write in normal mode", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "/project/src/foo.ts" }, { codexMode: false })
    );
    expect(result.decision).toBe("approve");
  });

  it("does not block Bash in Codex mode (R07 toolPattern is Write/Edit only)", () => {
    // R07's toolPattern is /^(?:Write|Edit|MultiEdit)$/ only
    // evaluateRules checks toolPattern so Bash does not match R07
    const result = evaluateRules(
      makeCtx("Bash", { command: "ls" }, { codexMode: true })
    );
    expect(result.decision).toBe("approve");
  });
});

// ============================================================
// R08: Breezing reviewer role guard
// ============================================================
describe("R08: Breezing reviewer role guard", () => {
  it("blocks Write for reviewer role", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "/project/src/foo.ts" }, { breezingRole: "reviewer" })
    );
    expect(result.decision).toBe("deny");
  });

  it("blocks Edit for reviewer role", () => {
    const result = evaluateRules(
      makeCtx("Edit", { file_path: "/project/src/foo.ts" }, { breezingRole: "reviewer" })
    );
    expect(result.decision).toBe("deny");
  });

  it("blocks git commit for reviewer role", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git commit -m 'test'" }, { breezingRole: "reviewer" })
    );
    expect(result.decision).toBe("deny");
  });

  it("blocks git push for reviewer role", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "git push origin main" }, { breezingRole: "reviewer" })
    );
    expect(result.decision).toBe("deny");
  });

  it("does not block ls for reviewer role (read-only command)", () => {
    const rule = GUARD_RULES.find((r) => r.id === "R08:breezing-reviewer-no-write")!;
    const result = rule.evaluate(
      makeCtx("Bash", { command: "ls -la" }, { breezingRole: "reviewer" })
    );
    // Does not match prohibited patterns, returns null
    expect(result).toBeNull();
  });

  it("does not block for non-reviewer role", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "/project/src/foo.ts" }, { breezingRole: "implementer" })
    );
    expect(result.decision).toBe("approve");
  });

  it("does not block when no role is set", () => {
    const result = evaluateRules(
      makeCtx("Write", { file_path: "/project/src/foo.ts" }, { breezingRole: null })
    );
    expect(result.decision).toBe("approve");
  });
});

// ============================================================
// R09: sensitive file Read warning (approve + systemMessage)
// ============================================================
describe("R09: sensitive file Read warning", () => {
  const secretPaths = [
    ".env",
    "id_rsa",
    "private.pem",
    "server.key",
    "secrets/api.json",
  ];

  for (const path of secretPaths) {
    it(`Read of ${path} returns approve + systemMessage`, () => {
      const result = evaluateRules(
        makeCtx("Read", { file_path: path })
      );
      expect(result.decision).toBe("approve");
      expect(result.systemMessage).toBeDefined();
      expect(result.systemMessage).toContain(path);
    });
  }

  it("Read of normal source file has no warning", () => {
    const result = evaluateRules(
      makeCtx("Read", { file_path: "src/index.ts" })
    );
    expect(result.decision).toBe("approve");
    expect(result.systemMessage).toBeUndefined();
  });
});

// ============================================================
// evaluateRules: integration tests
// ============================================================
describe("evaluateRules: integration tests", () => {
  it("skips rules when tool_input.command is not a string", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: 12345 })
    );
    expect(result.decision).toBe("approve");
  });

  it("returns approve when no rules match", () => {
    const result = evaluateRules(
      makeCtx("Bash", { command: "echo hello" })
    );
    expect(result.decision).toBe("approve");
  });

  it("Codex MCP tools are not subject to Bash rules", () => {
    // mcp__codex__* is not in GUARD_RULES (out of scope for rules.ts)
    // Blocked separately in index.ts
    const result = evaluateRules(
      makeCtx("mcp__codex__exec", { input: "ls" })
    );
    expect(result.decision).toBe("approve");
  });

  it("R01 takes priority when both R01 and R05 match", () => {
    // sudo + rm -rf command → R01 matches first
    const result = evaluateRules(
      makeCtx("Bash", { command: "sudo rm -rf /" })
    );
    expect(result.decision).toBe("deny");
    // deny includes R01's description
    expect(result.reason).toContain("sudo");
  });
});
